src/apps/config/alarm_codec.lua

```diff
@@ -0,0 +1,195 @@
+   end
+   function encoder:finish()
+      local size = 0
+      for _,src in ipairs(self.out) do size = size + ffi.sizeof(src) end
+      local dst = ffi.new('uint8_t[?]', size)
+      local pos = 0
+      for _,src in ipairs(self.out) do
+         ffi.copy(dst + pos, src, ffi.sizeof(src))
+         pos = pos + ffi.sizeof(src)
+      end
+      return dst, size
+   end
+   return encoder
+end
+
+function encode(alarm)
+   local name, args = unpack(alarm)
+   local codec = encoder()
+   codec:uint32(assert(alarm_codes[name], name))
+   return assert(alarms[name], name)(codec, unpack(args))
+end
+
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+local function decoder(buf, len)
+   local decoder = { buf=buf, len=len, pos=0 }
+   function decoder:read(count)
+      local ret = self.buf + self.pos
+      self.pos = self.pos + count
+      assert(self.pos <= self.len)
+      return ret
+   end
+   function decoder:uint32()
+      return ffi.cast(uint32_ptr_t, self:read(4))[0]
+   end
+   function decoder:string()
+      local len = self:uint32()
+      return ffi.string(self:read(len), len)
+   end
+   function decoder:finish(...)
+      return { ... }
+   end
+   return decoder
+end
+
+function decode(buf, len)
+   local codec = decoder(buf, len)
+   local name = assert(alarm_names[codec:uint32()])
+   return { name, assert(alarms[name], name)(codec) }
+end
+
+---
+
+local alarms_channel
+
+function get_channel()
+   if alarms_channel then return alarms_channel end
+   local name = '/'..S.getpid()..'/alarms-follower-channel'
+   local success, value = pcall(channel.open, name)
+   if success then
+      alarms_channel = value
+   end
+   return alarms_channel
+end
+
+local key_attrs = {'resource', 'alarm_type_id', 'alarm_type_qualifier'}
+local args_attrs = {'perceived_severity', 'alarm_text'}
+local function normalize (t, attrs)
+   t = t or {}
+   local ret = {}
+   for _, k in ipairs(attrs) do
+      table.insert(ret, t[k] or '')
+   end
+   return ret
+end
+local function normalize_key (t)
+   return normalize(t, key_attrs)
+end
+local function normalize_args (t)
+   return normalize(t, args_attrs)
+end
+
+-- To be used by the leader to group args into key and args.
+function parse_args (args)
+   local key = {
+      resource = args.resource,
+      alarm_type_id = args.alarm_type_id,
+      alarm_type_qualifier = args.alarm_type_qualifier,
+   }
+   local args = {
+      perceived_severity = args.perceived_severity,
+      alarm_text = args.alarm_text,
+   }
+   return key, args
+end
+
+function raise_alarm (key, args)
+   local channel = get_channel()
+   if channel then
+      local resource, alarm_type_id, alarm_type_qualifier = unpack(normalize_key(key))
+      local perceived_severity, alarm_text = unpack(normalize_args(args))
+      local buf, len = encode({'raise_alarm', {
+         resource, alarm_type_id, alarm_type_qualifier,
+         perceived_severity, alarm_text,
+      }})
+      channel:put_message(buf, len)
```

Comment:
I still haven't thought deeply about it. According to alarms spec a reasonable alarm rate is to raise/change an alarm every 5 minutes. Ideally a message should not be put in the queue if that won't result in a change in the alarms state. But that would require keep track of alarms state at this point and not the leader.

One possible solution is not to write messages directly to the channel but to place them in a table indexed by key. Then place a rate limiter in follower:pull that flushes this table to the channel every n milliseconds.